frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx

```diff
@@ -173,167 +174,230 @@ interface DashboardProps {
+    draftParameterValues,
+    editingOnLoad,
+    editingParameter,
+    fetchDashboard,
+    fetchDashboardCardData,
+    fetchDashboardCardMetadata,
+    initialize,
+    isAutoApplyFilters,
+    isEditing,
+    isFullscreen,
+    isHeaderVisible,
+    isNavigatingBackToDashboard,
+    isNightMode,
+    isSharing,
+    loadDashboardParams,
+    location,
+    onRefreshPeriodChange,
+    parameterValues,
+    parameters,
+    selectedTabId,
+    setDashboardAttributes,
+    setEditingDashboard,
+    setEditingParameter,
+    setErrorPage,
+    setParameterIndex,
+    setParameterValue,
+    setSharing,
+    toggleSidebar,
+  } = props;
+
+  const [isInitialized, setIsInitialized] = useState(false);
+  const [error, setError] = useState(null);
+  const [hasScroll, setHasScroll] = useState(getMainElement()?.scrollTop > 0);
+
+  const previousDashboard = usePrevious(dashboard);
+  const previousDashboardId = usePrevious(dashboardId);
+  const previousTabId = usePrevious(selectedTabId);
+  const previousParameterValues = usePrevious(parameterValues);
+
+  const visibleParameters = useMemo(
+    () => getVisibleParameters(parameters),
+    [parameters],
+  );
+
+  const tabHasCards = useMemo(() => {
+    if (!Array.isArray(dashboard?.dashcards)) {
+      return false;
     }
+    const tabDashCards = dashboard.dashcards.filter(
+      dc => dc.dashboard_tab_id === selectedTabId,
+    );
+    return tabDashCards.length > 0;
+  }, [dashboard, selectedTabId]);
+
+  const canWrite = Boolean(dashboard?.can_write);
+  const dashboardHasCards = dashboard?.dashcards.length > 0;
+  const hasVisibleParameters = visibleParameters.length > 0;
+
+  const shouldRenderAsNightMode = isNightMode && isFullscreen;
+
+  const shouldRenderParametersWidgetInViewMode =
+    !isEditing && !isFullscreen && hasVisibleParameters;
+
+  const shouldRenderParametersWidgetInEditMode =
+    isEditing && hasVisibleParameters;
+
+  const handleSetDashboardAttribute = useCallback(
+    <Key extends keyof IDashboard>(attribute: Key, value: IDashboard[Key]) => {
+      setDashboardAttributes({
+        id: dashboard.id,
+        attributes: { [attribute]: value },
+      });
+    },
+    [dashboard, setDashboardAttributes],
+  );
+
+  const handleSetEditing = useCallback(
+    (dashboard: IDashboard) => {
+      onRefreshPeriodChange(null);
+      setEditingDashboard(dashboard);
+    },
+    [onRefreshPeriodChange, setEditingDashboard],
+  );
+
+  const handleAddQuestion = useCallback(() => {
+    handleSetEditing(dashboard);
+    toggleSidebar(SIDEBAR_NAME.addQuestion);
+  }, [dashboard, toggleSidebar, handleSetEditing]);
+
+  const handleToggleSharing = useCallback(() => {
+    setSharing(!isSharing);
+  }, [isSharing, setSharing]);
+
+  const handleLoadDashboard = useCallback(
+    async (dashboardId: DashboardId) => {
+      initialize({ clearCache: !isNavigatingBackToDashboard });
+
+      loadDashboardParams();
+
+      const result: any = await fetchDashboard({
```

Comment:
I think we should be able to type `payload` in `fetchDashboard`'s result (`{ dashboard: IDashboard }`).